### PR TITLE
Added matrix-appservice-irc appservice resource

### PR DESCRIPTION
**Pull request by zwall-bp**
[matrix-appservice-irc](https://github.com/matrix-org/matrix-appservice-irc/tree/develop) is an appservice made by Matrix designed to bridge IRC. In the same way as our current Heisenbridge setup, but this appservice runs multiple IRC connections, per-user, talking from Matrix onto the IRC side.

Our plan is to add this to our current Matrix instance, and only allow it for private use. `osl-nodes` PR in the works.

### CODE REVIEW

File: libraries/helpers.rb [hunk 1]
```diff
@@ -1,6 +1,13 @@
 module OSLMatrix
   module Cookbook
     module Helpers
+      class ::Hash
```
Comment (ramereth) [inline libraries/helpers.rb:8]: Can you please add a comment for where this came from and what this does?

File: libraries/helpers.rb [hunk 2]
```diff
@@ -84,6 +91,70 @@ def osl_homeserver_db_defaults(pg_info)
...
+            'userTemplate' => '@as-irc_$NICK',
+          },
+          'ircClients' => {
+            'nickTemplate' => '$DISPLAY[OSL-Matrix]',
+            'kickOn' => {
+              'channelJoinFailure' => true,
+              'ircConnectionFailure' => true,
+              'userQuit' => true,
+            },
+          },
```
Comment (ramereth) [inline libraries/helpers.rb:158]: Is this just a place holder or required? Can you please this a bit more?

File: resources/matrix_irc.rb [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+property :key_appservice, String, default: lazy { osl_matrix_genkey(host_name + container_name) }
+property :key_homeserver, String, default: lazy { osl_matrix_genkey(container_name + host_name) }
+property :tag, String, default: 'latest'
+property :sensitive, [true, false], default: true
+
+action :create do
+  # Pull down the latest version
+  docker_image 'matrixdotorg/matrix-appservice-irc' do
+    tag new_resource.tag
+  end
```
Comment (ramereth) [inline resources/matrix_irc.rb:26]: When this image gets updated, how does the docker compose know that it needs to be updated? Have we tested this yet to make sure it works?

I believe in other cookbooks I've notified restarts when this gets triggered.

File: resources/matrix_irc.rb [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  #    volume "#{new_resource.host_path}/keys:/data"
+  #    entrypoint 'sh'
+  #    command ['-c', 'node lib/generate-signing-key.js > /data/signingkey.jwk && chmod 400 /data/signingkey.jwk']
+  #    user osl_synapse_user
+  #    not_if { ::File.exist?("#{new_resource.host_path}/keys/irc-signingkey.jwk") }
+  #  end
+  execute 'Generate signingkey' do
+    command "docker run --rm --entrypoint \"sh\" --volume #{new_resource.host_path}/keys:/data --user 994:989 matrixdotorg/matrix-appservice-irc \"-c\" \"node lib/generate-signing-key.js > /data/signingkey.jwk && chmod 400 /data/signingkey.jwk\""
+    creates "#{new_resource.host_path}/keys/signingkey.jwk"
+  end
```
Comment (ramereth) [inline resources/matrix_irc.rb:82]: Is this something we could generate manually and just have it in a databag item instead? Same for the passkey above?

File: libraries/helpers.rb (hunk 2 above)
Comment (zwall-bp) [reply libraries/helpers.rb:158]: This is the bare minimum config needed for every IRC server we add. I've made this to act as the starting template to begin the config with. All attributes can, and will be, modified by whatever is passed into the `config` property.

File: libraries/helpers.rb (hunk 1 above)
Comment (zwall-bp) [reply libraries/helpers.rb:8]: All configuration settings for anything yaml related are done programmicaly, with hashes. But I've came across a problem when trying to merge default settings with custom additions. Ruby doesn't come with a proper way of recursively merging Hashes, I'll add in an attribution and explanation for this addition.

File: resources/matrix_irc.rb (hunk 4 above)
Comment (zwall-bp) [reply resources/matrix_irc.rb:82]: I mean, we could. I was just trying to minimize the amount of static files/databags we needed to add in order to make these resources self-contained. These two resources are ran only one time, so it shouldn't be a worry for slowing down the scheduled converges.

As for the passkey, [`osl_hookshot` also runs the same command](https://github.com/osuosl-cookbooks/osl-matrix/blob/zwall/matrix-appservice-irc/resources/hookshot.rb#L138), so there is already precedent for that.

As for the signing key here, we do not utilize it at all, so it's also possible to hardcode a file into the program, if speed is an issue.

File: resources/matrix_irc.rb (hunk 3 above)
Comment (zwall-bp) [reply resources/matrix_irc.rb:26]: In the context of us running a Matrix server manage by `osl_synapse_service`, which we use in `osl-nodes`. As these appservices have been abstracted by being their own resource. When `docker_image` pulls a new image down, it should also trigger the parent `osl_matrix_irc`, which then [notifies docker to rebuild within `osl_synapse_service`](https://github.com/osuosl-cookbooks/osl-matrix/blob/zwall/matrix-appservice-irc/resources/synapse_service.rb#L77)

**Comment by zwall-bp** [comment]
I'm unsure of what your stance is on having a template config, which then adds on extra parameters that the developer passes when using the `config` property.

I was working with the idea that a resource should be able to produce a working, running container, even if no config was given.

File: resources/matrix_irc.rb (hunk 4 above)
Comment (ramereth) [reply resources/matrix_irc.rb:82]: Okay, you've documented this well enough that I think we can keep it in. If we run into issues moving forward, we can rethink that.

**Comment by ramereth** [comment]
> I'm unsure of what your stance is on having a template config, which then adds on extra parameters that the developer passes when using the `config` property.
> 
> I was working with the idea that a resource should be able to produce a working, running container, even if no config was given.

What do you mean? Maybe ping me in slack to clarify.

File: resources/matrix_irc.rb (hunk 3 above)
Comment (zwall-bp) [reply resources/matrix_irc.rb:26]: I've double checked this feature, and can confirm that `osl_dockercompose`'s `rebuild` action will successfully keep all `latest` tagged containers updated.

**Review by ramereth — APPROVED** [review]
minor bump

**Comment by zwall-bp** [comment]
!bump minor ~

**Comment by osuosl-manatee** [comment]
Jenkins has merged this PR into `master` and has automatically performed a minor-level version bump to v1.2.0.  Have a nice day!

**Pull request merged**